packages/apollo-codegen-typescript/src/codeGeneration.ts

```diff
@@ -379,10 +379,21 @@ export class TypescriptAPIGenerator extends TypescriptGenerator {
     }
 
     if (selection.selectionSet) {
-      return selection.selectionSet.selections.reduce(
-        this.reduceSelection,
-        acc
-      );
+      return selection.selectionSet.selections
+        .filter(childSelection => {
+          // Filter out unnecessary members from union types
+          if (
+            childSelection.kind === "TypeCondition" &&
+            selection.selectionSet &&
+            !selection.selectionSet.possibleTypes.some(
+              possibleType => possibleType.name === childSelection.type.name
+            )
+          ) {
+            return false;
+          }
+          return true;
+        })
```

Comment:
This feels a bit hacky to me and I'm pretty sure it doesn't fix other cases even though it fixes the test case. I'm working on fixing the other cases.

Reply:
I've updated the PR to fix the case in which there are nested fragments. I added a test case for it.